ironfish/src/wallet/wallet.ts

```diff
@@ -338,74 +338,62 @@ export class Wallet {
     accounts: ReadonlyArray<Account>,
   ): Promise<Map<string, Array<DecryptedNote>>> {
     const batchSize = 20
-    const notePromises: Array<
-      Promise<Array<{ accountId: string; decryptedNote: DecryptedNote }>>
-    > = []
+    const notePromises: Array<Promise<Map<string, Array<DecryptedNote | null>>>> = []
     let decryptNotesPayloads = []
-    for (const account of accounts) {
-      let currentNoteIndex = initialNoteIndex
-
-      for (const note of transaction.notes) {
-        decryptNotesPayloads.push({
-          accountId: account.id,
-          options: {
-            serializedNote: note.serialize(),
-            incomingViewKey: account.incomingViewKey,
-            outgoingViewKey: account.outgoingViewKey,
-            viewKey: account.viewKey,
-            currentNoteIndex,
-            decryptForSpender,
-          },
-        })
 
-        if (currentNoteIndex) {
-          currentNoteIndex++
-        }
+    let currentNoteIndex = initialNoteIndex
 
-        if (decryptNotesPayloads.length >= batchSize) {
-          notePromises.push(this.decryptNotesFromTransaction(decryptNotesPayloads))
-          decryptNotesPayloads = []
-        }
+    for (const note of transaction.notes) {
+      decryptNotesPayloads.push({
+        serializedNote: note.serialize(),
+        currentNoteIndex,
+        decryptForSpender,
+      })
+
+      if (currentNoteIndex) {
+        currentNoteIndex++
+      }
+
+      if (decryptNotesPayloads.length >= batchSize) {
+        notePromises.push(this.decryptNotesFromTransaction(accounts, decryptNotesPayloads))
```

Comment:
This might negatively affect performance with a large number of accounts because the parallelism will be limited. However, I have some changes planned that will significantly improve the performance, and this PR is a prerequisite for those changes. I also want to change the logic used by this function to improve the batching strategy.

Given that those changes are not ready yet, for I can tweak tweak this method to use an adjusted heuristic. Instead of using `batchSize` that counts the number of notes only, I can use a `workloadSize` that counts the number of notes times the number of accounts.